### DISCUSSION TITLE
Bug: interface methods that return type from same module break in Go SDK

### DISCUSSION THREAD

**Discussion by sipsma** [dagger/dagger#12335] [maintainers]
Repro: https://github.com/yann-soubeyrand/daggerverse/blob/0bfceb15f2597da89cb08bd6c8ba5c73baea6df7/test/main.go

The problem is that the autogenerated method for the interface impl tries to call `WithGraphQLQuery` but that doesn't exist on the type: https://github.com/yann-soubeyrand/daggerverse/blob/0bfceb15f2597da89cb08bd6c8ba5c73baea6df7/test/dagger.gen.go#L598

Due to the current rule that types defined in a module are serialized with their JSON representation rather than their ID, the fix isn't incredibly straightforward. The best I can think of right now is to call the interface method and get the ID of the object, but then look at the fields of the object and fill each of those in with individual API calls. That will result in a lot of API calls, but that's probably fine for now.
* An API for returning the JSON representation of a custom module object could help alleviate that, but would need to figure out the naming of the API (to minimize conflicts), whether to expose that in codegen or require it only be accessible from raw gql queries, etc.

**Comment by brakel** [comment]
Also having this issue trying to use interfaces with the Go SDK.

```go
type Authenticator interface {
	DaggerObject

	Authenticate(ctx context.Context, arns []string) (*AWSCredentials, error)
}

type AWSCredentials struct {
	AccessKeyID     *dagger.Secret
	SecretAccessKey *dagger.Secret
	SessionToken    *dagger.Secret
}
```

```go
func (r *authenticatorImpl) Authenticate(ctx context.Context, arns []string) (*AWSCredentials, error) {
	q := r.query.Select("authenticate")
	q = q.Arg("arns", arns)

	return (&AWSCredentials{}).WithGraphQLQuery(q)
}

func (r *authenticatorImpl) WithGraphQLQuery(q *querybuilder.Selection) Authenticator {
	return &authenticatorImpl{query: q}
}
```
